crates/op-rbuilder/src/primitives/reth/engine_api_builder.rs

```diff
@@ -0,0 +1,350 @@
+use reth_node_core::version::{CARGO_PKG_VERSION, CLIENT_CODE, VERGEN_GIT_SHA};
+use reth_optimism_node::OP_NAME_CLIENT;
+pub use reth_optimism_rpc::OpEngineApi;
+use reth_optimism_rpc::{engine::OP_ENGINE_CAPABILITIES, OpEngineApiServer};
+use reth_payload_builder::PayloadStore;
+use reth_rpc_engine_api::EngineCapabilities;
+
+use alloy_eips::eip7685::Requests;
+use alloy_primitives::{BlockHash, B256, U64};
+use alloy_rpc_types_engine::{
+    ClientVersionV1, ExecutionPayloadBodiesV1, ExecutionPayloadInputV2, ExecutionPayloadV3,
+    ForkchoiceState, ForkchoiceUpdated, JwtSecret, PayloadId, PayloadStatus,
+};
+use jsonrpsee_core::{client::SubscriptionClientT, server::RpcModule, RpcResult};
+use op_alloy_rpc_types_engine::{
+    OpExecutionData, OpExecutionPayloadV4, ProtocolVersion, SuperchainSignal,
+};
+use reth_chainspec::EthereumHardforks;
+use reth_node_api::{EngineTypes, EngineValidator};
+use reth_rpc_api::IntoEngineApiRpcModule;
+use reth_rpc_layer::AuthClientLayer;
+use reth_storage_api::{BlockReader, HeaderProvider, StateProviderFactory};
+use reth_tasks::TaskExecutor;
+use reth_transaction_pool::TransactionPool;
+use tracing::{self, log::warn};
+use url::Url;
+
+/// Configuration for an engine peer with JWT authentication
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct EnginePeer {
+    pub url: Url,
+    pub jwt: JwtSecret,
+}
+
+impl EnginePeer {
+    pub fn new(url: Url, jwt_path: JwtSecret) -> Self {
+        Self { url, jwt: jwt_path }
+    }
+
+    pub fn http_client(&self) -> impl SubscriptionClientT + Clone + Send + Sync + Unpin + 'static {
+        // Create a middleware that adds a new JWT token to every request.
+        let secret_layer = AuthClientLayer::new(self.jwt);
+        let middleware = tower::ServiceBuilder::default().layer(secret_layer);
+        jsonrpsee::http_client::HttpClientBuilder::default()
+            .set_http_middleware(middleware)
+            .build(self.url.clone())
+            .expect("Failed to create http client")
+    }
+}
+
+/// Builder for basic [`OpEngineApi`] implementation.
+#[derive(Debug)]
+pub struct OpEngineApiBuilder<EV> {
+    engine_validator_builder: EV,
+    engine_peers: Vec<EnginePeer>,
+}
+
+impl<EV> OpEngineApiBuilder<EV> {
+    /// Create a new builder with engine peers configuration
+    pub fn with_engine_peers(mut self, engine_peers: Vec<EnginePeer>) -> Self {
+        self.engine_peers = engine_peers;
+        self
+    }
+}
+
+impl<EV> Default for OpEngineApiBuilder<EV>
+where
+    EV: Default,
+{
+    fn default() -> Self {
+        Self {
+            engine_validator_builder: EV::default(),
+            engine_peers: Vec::new(),
+        }
+    }
+}
+
+impl<N, EV> EngineApiBuilder<N> for OpEngineApiBuilder<EV>
+where
+    N: FullNodeComponents<
+        Types: NodeTypes<
+            ChainSpec: EthereumHardforks,
+            Payload: EngineTypes<ExecutionData = OpExecutionData>,
+        >,
+    >,
```

Comment:
We have now `NodeBounds` in `traits.rs`. This block of code can be simplified into: `N: NodeBounds`.